geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java

```diff
@@ -2960,24 +2964,15 @@ private boolean readHandshakeForReceiver(final DataInput dis) {
       final boolean isSecure = authInit != null && !authInit.isEmpty();
 
       if (isSecure) {
-        if (owner.getConduit().waitForMembershipCheck(remoteMember)) {
-          sendOKHandshakeReply();
-          notifyHandshakeWaiter(true);
-        } else {
-          // check if we need notifyHandshakeWaiter() call.
+        if (!owner.getConduit().waitForMembershipCheck(remoteMember)) {
           notifyHandshakeWaiter(false);
-          logger.warn("{} timed out during a membership check.",
-              p2pReaderName());
+          logger.warn("{} timed out during a membership check.", p2pReaderName());
+          requestClose("timed out during a membership check");
           return true;
         }
-      } else {
-        sendOKHandshakeReply();
-        try {
-          notifyHandshakeWaiter(true);
-        } catch (Exception e) {
-          logger.fatal("Uncaught exception from listener", e);
-        }
       }
+      sendOKHandshakeReply();
+      notifyHandshakeWaiter(true);
```

Comment:
The logic simplifications in this method make the flow easier to understand.

The only functional change I see (besides the new `requestClose()`, is that if `notifyHandshakeWaiter()` throws an `Exception` we won't log a fatal-level message. That catch block was introduced by:

GEODE-2113 Implement SSL over NIO 

I perused `notifyHandshakeWaiter()` and the methods it calls and I don't see any possibility of unhandled exceptions. As a result I think it's ok to remove that catch block.